src/overlays/actors/ovl_Elf_Msg4/z_elf_msg4.c

```diff
@@ -30,30 +33,137 @@ const ActorInit Elf_Msg4_InitVars = {
+            return true;
+        }
+    }
+    if (ELFMSG4_GET_7F00(&this->actor) == 0x7F) {
+        return false;
+    }
+    if (Flags_GetSwitch(play, ELFMSG4_GET_7F00(&this->actor)) != 0) {
+        Actor_MarkForDeath(&this->actor);
+        return true;
+    }
+    return false;
+}
+
+void ElfMsg4_Init(Actor* thisx, PlayState* play) {
+    ElfMsg4* this = THIS;
+
+    if (func_80AFD380(this, play) == 0) {
+        Actor_ProcessInitChain(&this->actor, sInitChain);
+
+        if (ABS_ALT(this->actor.home.rot.x) == 0) {
+            this->actor.scale.z = 0.4f;
+            this->actor.scale.x = 0.4f;
+        } else {
+            this->actor.scale.x = this->actor.scale.z = ABS_ALT(this->actor.home.rot.x) * 0.04f;
+        }
+        if (this->actor.home.rot.z == 0) {
+            this->actor.scale.y = 0.4f;
+        } else {
+            this->actor.scale.y = this->actor.home.rot.z * 0.04f;
+        }
+        this->actor.shape.rot.z = 0;
+        this->actionFunc = func_80AFD770;
+        this->unk_144 = 0;
+        this->actor.shape.rot.x = this->actor.shape.rot.y = this->actor.shape.rot.z;
+    }
+}
+
+void ElfMsg4_Destroy(Actor* thisx, PlayState* play) {
+}
+
+s32 func_80AFD5B4(ElfMsg4* this) {
+    if (ELFMSG4_GET_8000(&this->actor) != 0) {
+        return ELFMSG4_GET_FF(&this->actor) + 0x200;
+    } else {
+        return -0x200 - ELFMSG4_GET_FF(&this->actor);
+    }
+}
+
+s32 func_80AFD5E0(ElfMsg4* this) {
+    return (this->actor.xzDistToPlayer < (100.0f * this->actor.scale.x)) && (this->actor.playerHeightRel >= 0.0f) &&
+           (this->actor.playerHeightRel < (100.0f * this->actor.scale.y));
+}
+
+void func_80AFD668(ElfMsg4* this, PlayState* play) {
+    Player* player = (Player*)play->actorCtx.actorLists[ACTORCAT_PLAYER].first;
+    EnElf* tatl = (EnElf*)player->tatlActor;
+
+    if ((player->tatlActor != NULL) && (func_80AFD5E0(this))) {
+        player->tatlTextId = func_80AFD5B4(this);
+        ActorCutscene_SetIntentToPlay(0x7C);
+        tatl->elfMsg = this->unk_144;
+        if (this->actor.cutscene == -1) {
+            this->actor.cutscene = 0x7C;
+        }
+        if ((player->tatlTextId < 0) && (this->actor.home.rot.x < 0)) {
+            if (ActorCutscene_GetCurrentIndex() == 0x7D) {
+                ActorCutscene_Stop(0x7D);
+                ActorCutscene_SetIntentToPlay((s16)this->actor.cutscene);
+            } else if (ActorCutscene_GetCanPlayNext((s16)this->actor.cutscene)) {
+                ActorCutscene_Start(this->actor.cutscene, &this->actor);
+                func_800E0348(play->cameraPtrs[0]);
+            } else {
+                ActorCutscene_SetIntentToPlay((s16)this->actor.cutscene);
+            }
+        }
+    }
+}
+
+void func_80AFD770(ElfMsg4* this, PlayState* play) {
+    Actor* actorPtr = play->actorCtx.actorLists[ACTORCAT_BG].first;
+
+    while (actorPtr != NULL) {
+        if ((actorPtr->id != 0x1D8) || (ELFMSG4_GET_FF(&this->actor) != (actorPtr->params & 0xFF)) ||
```

Comment:
```suggestion
        if ((actorPtr->id != ACTOR_ELF_MSG5) || (ELFMSG4_GET_FF(&this->actor) != ELFMSG5_GET_FF(actorPtr)) ||
```
Based on the `id` check, we know that `ActorPtr` should be an `ElfMsg5`